ci/check-nix-tools.nix

```diff
@@ -14,21 +12,68 @@ writeScript "check-nix-tools.sh" ''
 
   set -euo pipefail
 
-  export PATH="${makeBinPath [ stdenv.shellPackage coreutils nixStable git ]}:$PATH"
+  export PATH="${makeBinPath [ stdenv.shellPackage coreutils nixStable git gawk ]}:$PATH"
 
   cd $(git rev-parse --show-toplevel)
 
-  fail_nix_tools_check() {
-    git diff -w --text > /tmp/nix-tools.patch
-    if [ -n "''${BUILDKITE_JOB_ID:-}" ]; then
-      buildkite-agent artifact upload /tmp/nix-tools.patch --job "$BUILDKITE_JOB_ID"
-    fi
-    echo "ERROR: you need to (run ./nix/regenerate.sh or apply the patch in the buildkite artifact) and commit the changes" >&2
-    exit 1
-  }
-
   # The regenerate script is here by convention
   ./nix/regenerate.sh
 
-  git diff -w --text --exit-code || fail_nix_tools_check
+  # The generated files will appear somewhere under ./nix
+  git add -A nix
+
+  # Check if there are changes staged for commit.
+  if git diff-index --cached --quiet HEAD --; then
+    echo "Generated Nix code is up-to-date."
+    exit 0
+  else
+    echo "Committing changes..."
+    commit_message="Regenerate nix"
+
+    # If on a PR branch, search for a previous regen commit to fix up.
+    commit_fixup=""
+    if [ -n "''${BUILDKITE_PULL_REQUEST_BASE_BRANCH:-}" ]; then
+      git fetch -v origin $BUILDKITE_PULL_REQUEST_BASE_BRANCH
+      commit_fixup=$(git log --pretty=oneline --no-decorate origin/$BUILDKITE_PULL_REQUEST_BASE_BRANCH..HEAD | awk "/$commit_message/ { print \$1; }")
+    fi
+
+    # Create the commit
+    export GIT_COMMITTER_NAME="IOHK"
+    export GIT_COMMITTER_EMAIL="devops+nix-tools@iohk.io"
+    export GIT_AUTHOR_NAME="$GIT_COMMITTER_NAME"
+    export GIT_AUTHOR_EMAIL="$GIT_COMMITTER_EMAIL"
+    if [ -n "$commit_fixup" ]; then
+      git commit --no-gpg-sign --fixup "$commit_fixup"
```

Comment:
Since this lead to force pushes, I am not too sure I am in favor of it. This will inevitably breaks some developers workflows. At the minimum I would prefer we only fixup with the last commit but not further ancient commit.

Reply:
[`git commit --fixup`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupltcommitgt) will not overwrite git history. It adds a new commit. Then it's up to the user whether they would like to squash with [`git rebase --autosquash`](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---autosquash). This is quite cool I think.

Reply:
ah very nice indeed!